File: 研究/Scripts/stars.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;
using System.Text;


public class Stars
{
    static int star_count = 2500;
    static public void build_stars(Mesh mesh)
    {
        Vector3[] star_vertex = new Vector3[star_count];
        Vector2[] star_size = new Vector2[star_count];
        int[] star_index = new int[star_count];
        for (int i = 0; i < star_count; ++i)
        {
            Vector3 v = Random.insideUnitSphere;
            v.Normalize();
            v *= 50f;
            star_vertex[i] = v;//.Set(v.x, v.y, v.z, Mathf.Lerp(0.1f, 2.5f, Mathf.Pow(Random.value, 10f)));
            star_size[i].x = Mathf.Lerp(0.1f, 2.5f, Mathf.Pow(Random.value, 10f));
            star_index[i] = i;
        }

        mesh.vertices = star_vertex;
        mesh.SetIndices(star_index, MeshTopology.Points, 0);
        mesh.uv = star_size;

//         for (var e = vec3.create(), r = new Float32Array(t << 2), n = 0; n < r.length; n += 4) Random.unitVec3(e),
// 				vec3.scale(e, e, 50),
// 				r[n + 0] = e[0],
// 				r[n + 1] = e[1],
// 				r[n + 2] = e[2],
// 				r[n + 3] = lerp(0.1, 2.5, Math.pow(Math.random(), 10));
    }
}

File: three/stars.js
/**
 * Created by Administrator on 2016/5/5.
 */
define(['shaders'], function(shaders){
    var stars;
    var star_count=2500;
    var radius = 20;
    var geometry=new THREE.BufferGeometry();


    var positions=new THREE.BufferAttribute(new Float32Array(star_count*3), 3);
    var star_size=new THREE.BufferAttribute(new Float32Array(star_count*2), 2);
    var star_index=new Uint32Array(star_count);


    var point=new THREE.Vector3();
    for(var i=0; i<star_count; i++){
        var v=new THREE.Vector3();
        /* v=Random.insideUnitSphere;
         v.Normalize();
         v*=50;
         point[i]=v;
         */

        var px=(Math.random()*100-50)*20;
        var py=(Math.random()*100-50)*20;
        var pz=(Math.random()*100-50)*20;
        positions.setXYZ(i, px, py, pz);


        /*var x=lerp(0.1, 2.5, Math.pow(Math.random(), 10));


         star_size.setX(i, x)*/
        star_index[i]=i;

    }


    geometry.setIndex(new THREE.BufferAttribute(star_index, 1));
    geometry.addAttribute('position', positions);
    geometry.addAttribute('uv', star_size);
    geometry.computeBoundingSphere();

/*    var materials=new THREE.PointsMaterial({color: new THREE.Color( 0xffffff )});*/

    var customUniforms=shaders.shader['stars'].uniforms;
    var materials=new THREE.ShaderMaterial(
        {
            uniforms: customUniforms,
            vertexShader: shaders.shader['stars'].vertexShader,
            fragmentShader: shaders.shader['stars'].fragmentShader
        });

    stars=new THREE.Points(geometry, materials);


    console.log("stars.js");
    return {
        stars: stars
    }

});
